src/compiler/compiler.rs

```diff
@@ -556,41 +560,44 @@ gcc
             //TODO: do something smarter here.
             if line == "gcc" {
                 debug!("Found GCC");
-                return future::ok(Some(CompilerKind::Gcc)).boxed()
+                return future::ok(Some(Box::new(CCompiler(GCC)) as Box<Compiler<T>>)).boxed()
             } else if line == "clang" {
                 debug!("Found clang");
-                return future::ok(Some(CompilerKind::Clang)).boxed()
+                return future::ok(Some(Box::new(CCompiler(Clang)) as Box<Compiler<T>>)).boxed()
             } else if line == "msvc" {
                 debug!("Found MSVC");
                 let prefix = msvc::detect_showincludes_prefix(&creator,
                                                               executable.as_ref(),
                                                               &pool);
                 return Box::new(prefix.map(|prefix| {
                     trace!("showIncludes prefix: '{}'", prefix);
-                    Some(CompilerKind::Msvc {
+                    Some(Box::new(CCompiler(MSVC {
                         includes_prefix: prefix,
-                    })
+                    })) as Box<Compiler<T>>)
                 }))
             }
         }
         future::ok(None).boxed()
     }))
 }
 
-/// If `executable` is a known compiler, return `Some(Compiler)` containing information about it.
+/// If `executable` is a known compiler, return a `CompilerInfo` containing information about it.
 pub fn get_compiler_info<T>(creator: &T, executable: &str, pool: &CpuPool)
-                            -> SFuture<Compiler>
+                            -> SFuture<CompilerInfo<T>>
     where T: CommandCreatorSync
 {
     let executable = executable.to_string();
     let pool = pool.clone();
-    Box::new(detect_compiler_kind(creator, &executable, &pool).and_then(move |kind| {
-        match kind {
-            Some(kind) => {
-                pool.spawn_fn(move || Compiler::new(&executable, kind))
-                    .chain_err(|| "failed to learn compiler metadata")
+    let detect = detect_compiler(creator, &executable, &pool);
+    Box::new(detect.and_then(move |compiler| {
+        match compiler {
+            Some(compiler) => {
+                Box::new(sha1_digest(executable.clone(), &pool)
```

Comment:
That's a really good point and I can't believe I didn't realize it! I don't have a non-rustup install handy, but would it be enough to run `rustc --print sysroot` and hash `$sysroot/bin/rustc`? I'd have to sanity check how that works with the tooltool-downloaded rustc we use in Firefox CI.

We discussed the dylib issue in Hawaii and we kind of hand-waved it away. I definitely don't want to be parsing the output of `ldd` or reading the binary headers to figure this out, but maybe it's enough to hash rustc + everything in `$sysroot/lib/`? Actually given that rustc is just a stub, maybe we skip hashing rustc entirely, hash `$sysroot/lib` and that's good enough?